### PR TITLE
Issue 3564 var override

### PR DESCRIPTION
**Pull request by ShubhamRasal**
- previously, we were checking any function/expression in variable
- now, update the logic, lazy eval only if variable contains any
  protocol variable(global)
- override the variable value if passed from cli

---

- [x] Test to cover https://github.com/projectdiscovery/nuclei/issues/3580

### CODE REVIEW

**Comment by ShubhamRasal** [comment]
already updated the integration tests

**Review by ehsandeep — CHANGES_REQUESTED** [review]
Add test to cover - https://github.com/projectdiscovery/nuclei/issues/3580

**Review by ehsandeep — APPROVED** [review]

File: v2/pkg/protocols/common/variables/variables.go [hunk 1]
```diff
@@ -8,11 +8,13 @@ import (
 	"github.com/projectdiscovery/nuclei/v2/pkg/protocols/common/expressions"
 	"github.com/projectdiscovery/nuclei/v2/pkg/protocols/common/generators"
 	"github.com/projectdiscovery/nuclei/v2/pkg/protocols/common/interactsh"
-	"github.com/projectdiscovery/nuclei/v2/pkg/protocols/common/marker"
 	"github.com/projectdiscovery/nuclei/v2/pkg/types"
 	"github.com/projectdiscovery/nuclei/v2/pkg/utils"
+	"github.com/projectdiscovery/stringsutil"
```
Comment (tarunKoyalwar) [inline v2/pkg/protocols/common/variables/variables.go:13]: @ShubhamRasal we should replace this with utils/strings

File: v2/pkg/protocols/common/variables/variables.go [hunk 2]
```diff
@@ -8,11 +8,13 @@ import (
...
 )
 
+var DefaultProtocolVariables = []string{"FQDN", "RDN", "DN", "TLD", "SD", "BaseURL", "RootURL", "Hostname", "Host", "Port", "Path", "File", "Scheme", "Input"}
```
Comment (tarunKoyalwar) [inline v2/pkg/protocols/common/variables/variables.go:16]: @ShubhamRasal I think we can revisit lazyVar implementation . 
Keeping track of whitelist does not seem like best solution . we can try some other method if possible 

cc: @Mzack9999

**Review by tarunKoyalwar — CHANGES_REQUESTED** [review]

File: v2/pkg/protocols/common/variables/variables.go (hunk 2 above)
Comment (Mzack9999) [reply v2/pkg/protocols/common/variables/variables.go:16]: A few alternatives:
- static map/helper function in each protocol with known variable names
- As noted by @tarunKoyalwar, the to map shared conversion functions (http, dns) are in the utils package (as they are used by multiple tools). Not sure if centralizing the logic might be an option

**Review by tarunKoyalwar — APPROVED** [review]
lgtm !

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**